### PR TITLE
Use akka management for cluster bootstrapping

### PR DESCRIPTION
**Pull request by longshorej**
This removes the registrar-based cluster bootstrapping implementation in favor of one that uses `akka-management`. There's not too much here other than to bind it to our own service locator instead of the provided one, and configure some settings according to our own conventions.

You'll have to publish this PR locally for this to work: https://github.com/akka/akka-management/pull/58

### CODE REVIEW

File: akka-cluster-bootstrap/src/main/resources/application.conf [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# TODO: What needs to happen to move this into rp.conf?
+
+akka {
```
Comment (patriknw) [inline akka-cluster-bootstrap/src/main/resources/application.conf]: How is this application.conf supposed to be used? This is a library, right? Supposed to be included as a dependency in the application. A library can't include an application.conf in its jar. That would be a high risk classpath conflict/duplicate. A library can't override properties owned by another library in a reference.conf. That means that what you are trying to do is not possible and should instead be defined by the end user in their application.conf, possibly by including rp.conf.

In short, you can't and shouldn't try to make it automagic. Something must be added in the application's conf to enable this.

One exception is that Play has the special reference-overrides.conf, but that can only be used with Play and Lagom.

File: akka-cluster-bootstrap/src/main/resources/application.conf [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  cluster {
+    http.management {
+      hostname = ${?RP_ENDPOINT_AKKA_MGMT_HTTP_BIND_HOST}
+      port = ${?RP_ENDPOINT_AKKA_MGMT_HTTP_BIND_PORT}
+    }
+  }
+
+  remote {
+    # akka-remote => AKKA_REMOTE
+    # AKKA_REMOTE => akka-remote
```
Comment (ktoso) [inline akka-cluster-bootstrap/src/main/resources/application.conf]: ?

File: akka-cluster-bootstrap/src/main/scala/com/lightbend/rp/akkaclusterbootstrap/ClusterServiceDiscovery.scala [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    ServiceLocator
+      .lookup(name, "akka-mgmt-http")(system)
+      .map(services =>
+        Resolved(
+          name,
+          services
+            .filter(s => s.uri.getHost != null && s.uri.getPort > 0)
+            .map(s => ResolvedTarget(s.uri.getHost, Some(s.uri.getPort)))))
+  }
+}
```
Comment (ktoso) [inline akka-cluster-bootstrap/src/main/scala/com/lightbend/rp/akkaclusterbootstrap/ClusterServiceDiscovery.scala]: be aware through that that API is not set in stone yet; once/if you ship this, will we be able to change things?

File: build.sbt [hunk 4]
```diff
@@ -2,19 +2,18 @@ import scala.collection.immutable.Seq
...
-  val play25            = "2.5.0"
-  val play26            = "2.6.0"
-  val scala211          = "2.11.11"
-  val scala212          = "2.12.3"
-  val scalaJava8Compat  = "0.8.0"
-  val scalaTest         = "3.0.1"
-  val sprayJson         = "1.3.3"
-  val typesafeConfig    = "1.3.1"
+  val akka                      = "2.4.12" // First version cross-compiled to 2.12
+  val akkaDns                   = "2.4.2"
```
Comment (ktoso) [inline build.sbt]: I noticed that akka dns isn't released to maven central btw, is this a problem for users of this lib or who uses it?

File: akka-cluster-bootstrap/src/main/resources/rp.conf [hunk 5]
```diff
@@ -1,23 +1,27 @@
-# This needs to be in application.conf because reference.conf files have no guarantee
-# and Akka's values may override ours. If we were strictly in a Play app, we could use
-# reference-overrides.conf but as Akka Cluster also can be used standalone, here we are.
-
 akka {
```
Comment (patriknw) [inline akka-cluster-bootstrap/src/main/resources/rp.conf]: I'd give this file a better name, such as `reactive-lib-defaults.conf`, and describe how it's intended to be used (I guess it's intended to be included from the application's conf.

File: akka-cluster-bootstrap/src/main/resources/rp.conf [hunk 6]
```diff
@@ -1,23 +1,27 @@
...
-      port = ${?RP_ENDPOINT_AKKA_REMOTE_HOST}
+      resolve-interval = 1 second
 
-      bind-hostname = ${?RP_ENDPOINT_AKKA_REMOTE_BIND_HOST}
-      bind-port = ${?RP_ENDPOINT_AKKA_REMOTE_BIND_HOST}
+      resolve-timeout = 2 second
     }
   }
+
+  extensions += "com.lightbend.rp.akkaclusterbootstrap.ClusterBootstrap"
```
Comment (patriknw) [inline akka-cluster-bootstrap/src/main/resources/rp.conf]: Perhaps this works when defined in application.conf (rp.conf included in application.conf). Anyway, we had trouble with that when trying to auto enable Akka Diagnostics from reference.conf and therefore `akka.library-extensions` was added:
https://github.com/lightbend/akka-commercial-addons/blob/master/akka-diagnostics/src/main/resources/reference.conf#L11

File: build.sbt (hunk 4 above)
Comment (longshorej) [reply build.sbt]: Ideally I'd like to reorganize the akka-management project a bit as I proposed in email (having a core subproject with minimal dependencies, marked as provided) and provide this as a shaded dependency inside of reactive-lib.

File: akka-cluster-bootstrap/src/main/scala/com/lightbend/rp/akkaclusterbootstrap/ClusterServiceDiscovery.scala (hunk 3 above)
Comment (longshorej) [reply akka-cluster-bootstrap/src/main/scala/com/lightbend/rp/akkaclusterbootstrap/ClusterServiceDiscovery.scala]: Sure, we can always change usage as necessary.

File: akka-cluster-bootstrap/src/main/resources/application.conf (hunk 1 above)
Comment (longshorej) [reply akka-cluster-bootstrap/src/main/resources/application.conf]: I understand your concern, hence this part being marked TODO. What we'll probably end up doing is documenting that you include `rp.conf` in your `application.conf` file for regular apps. For Play and Lagom we can leverage `reference-overrides.conf`.

File: akka-cluster-bootstrap/src/main/resources/application.conf (hunk 2 above)
Comment (longshorej) [reply akka-cluster-bootstrap/src/main/resources/application.conf]: That shouldn't be there :p

**Comment by fsat** [comment]
There's the following [build failure](https://travis-ci.org/lightbend/reactive-lib/builds/308572725#L631):

```
[error] sbt.librarymanagement.ResolveException: unresolved dependency: com.lightbend.akka#akka-management-cluster-http_2.12;0.4+9-2f03f4eb: not found
```

I'll restart the build since `akka-management-cluster-http` for Scala `2.12` is available at version `0.5`.

**Comment by ktoso** [comment]
Yah but 0.5 does not include the discovery things. See what happened since 0.5 here: https://github.com/akka/akka-management/compare/v0.5...master

**Comment by longshorej** [comment]
I've updated the build to publish the current 2.4 locally.
